### PR TITLE
2283-delete Category endpoint and unit tests

### PR DESCRIPTION
**Pull request by Akrish17**
[2283- Delete Category endpoint and unit tests]

#### 📲 What

Implemented a new Rest endpoint for deleting a category from a menu

#### 🤔 Why

Can only delete a category using this endpoint

#### 🛠 How

Pass the menuId where the category exists and Category Id to be deleted in the request

#### 👀 Evidence

Screenshots / external resources / links / etc.
Link to documentation updated with changes impacted in the PR

image

#### 🕵️ How to test

Notes for QA
Swagger JSON is available here: http://localhost:9000/swagger/oas.json
UI Swagger here: http://localhost:9000/swagger/index.html
#### ✅ Acceptance criteria Checklist

- [ ] Code peer reviewed?
- [ ] Documentation has been updated to reflect the changes?
- [ ] Passing all automated tests, including a successful deployment?
- [ ] Passing any exploratory testing?
- [ ] Rebased/merged with latest changes from development and re-tested?
- [ ] Meeting the Coding Standards?

### CODE REVIEW

**Review by ekartsonakisamido — APPROVED** [review]

File: java/src/main/java/com/xxAMIDOxx/xxSTACKSxx/menu/api/v1/DeleteCategoryController.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    @DeleteMapping(consumes = "application/json", produces = "application/json")
+    @Operation(
+            tags = "Category",
+            summary = "Update a category in the menu",
+            description = "Update a category to menu",
+            operationId = "UpdateMenuCategory",
+            responses = {
+                    @ApiResponse(
+                            responseCode = "200",
+                            description = "Resource created",
```
Comment (peet-j) [inline java/src/main/java/com/xxAMIDOxx/xxSTACKSxx/menu/api/v1/DeleteCategoryController.java:33]: created?

File: java/src/main/java/com/xxAMIDOxx/xxSTACKSxx/menu/api/v1/DeleteCategoryController.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+                            content = @Content(mediaType = "application/json",
+                                    schema = @Schema(implementation = ResourceCreatedResponse.class))),
+                    @ApiResponse(
```
Comment (peet-j) [inline java/src/main/java/com/xxAMIDOxx/xxSTACKSxx/menu/api/v1/DeleteCategoryController.java:36]: The response schema type doesn't look correct either

File: java/src/main/java/com/xxAMIDOxx/xxSTACKSxx/menu/api/v1/UpdateCategoryController.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    @PutMapping(consumes = "application/json", produces = "application/json")
+    @Operation(
+            tags = "Category",
+            summary = "Update a category in the menu",
+            description = "Update a category to menu",
+            operationId = "UpdateMenuCategory",
+            responses = {
+                    @ApiResponse(
+                            responseCode = "200",
+                            description = "Resource created",
```
Comment (peet-j) [inline java/src/main/java/com/xxAMIDOxx/xxSTACKSxx/menu/api/v1/UpdateCategoryController.java:36]: again

File: java/src/main/java/com/xxAMIDOxx/xxSTACKSxx/menu/api/v1/UpdateCategoryController.java [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+                            content = @Content(mediaType = "application/json",
+                                    schema = @Schema(implementation = ResourceCreatedResponse.class))),
+                    @ApiResponse(
```
Comment (peet-j) [inline java/src/main/java/com/xxAMIDOxx/xxSTACKSxx/menu/api/v1/UpdateCategoryController.java:39]: and again

File: java/src/main/java/com/xxAMIDOxx/xxSTACKSxx/menu/api/v1/impl/DeleteCategoryControllerImpl.java [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+    private DeleteCategoryHandler handler;
+
+    public DeleteCategoryControllerImpl(DeleteCategoryHandler handler) {
+        this.handler = handler;
+    }
+
+    @Override
+    public ResponseEntity<Void> deleteCategory(UUID menuId, UUID categoryId,
+                                               String correlationId) {
+        DeleteCategoryCommand command = map(correlationId, menuId, categoryId);
```
Comment (peet-j) [inline java/src/main/java/com/xxAMIDOxx/xxSTACKSxx/menu/api/v1/impl/DeleteCategoryControllerImpl.java:26]: covered in other PR

File: java/src/main/java/com/xxAMIDOxx/xxSTACKSxx/menu/api/v1/impl/DeleteMenuControllerImpl.java [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+    private DeleteMenuHandler handler;
+
+    public DeleteMenuControllerImpl(DeleteMenuHandler handler) {
+        this.handler = handler;
+    }
+
+    @Override
+    public ResponseEntity<Void> deleteMenu(UUID menuId, String correlationId) {
+
+        DeleteMenuCommand command = map(correlationId, menuId);
```
Comment (peet-j) [inline java/src/main/java/com/xxAMIDOxx/xxSTACKSxx/menu/api/v1/impl/DeleteMenuControllerImpl.java:31]: Again.. pointless mapping. Just instantiate the command.

File: java/src/main/java/com/xxAMIDOxx/xxSTACKSxx/menu/api/v1/impl/DeleteMenuControllerImpl.java [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+        handler.handle(command);
+        return new ResponseEntity<>(OK);
```
Comment (peet-j) [inline java/src/main/java/com/xxAMIDOxx/xxSTACKSxx/menu/api/v1/impl/DeleteMenuControllerImpl.java:33]: Nothing in a successful response. Prime candidate for a 204 no?

File: java/src/main/java/com/xxAMIDOxx/xxSTACKSxx/menu/handlers/DeleteCategoryHandler.java [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+    }
+
+    @Override
+    List<MenuEvent> raiseApplicationEvents(Menu menu,
+                                           DeleteCategoryCommand command) {
+        return Arrays.asList(new MenuUpdatedEvent(command),
+                new CategoryDeletedEvent(command, command.getCategoryId()));
+    }
+
+    Category getCategory(Menu menu, DeleteCategoryCommand command) {
```
Comment (peet-j) [inline java/src/main/java/com/xxAMIDOxx/xxSTACKSxx/menu/handlers/DeleteCategoryHandler.java:54]: You're likely doing the same check in UpdateCategory as DeleteCategory - try to avoid duplicating code

File: java/src/main/java/com/xxAMIDOxx/xxSTACKSxx/menu/handlers/UpdateCategoryHandler.java [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+    }
+
+    @Override
+    List<MenuEvent> raiseApplicationEvents(Menu menu,
+                                           UpdateCategoryCommand command) {
+        return Arrays.asList(new MenuUpdatedEvent(command),
+                new CategoryUpdatedEvent(command, command.getCategoryId()));
+    }
+
+    Category getCategory(Menu menu, UpdateCategoryCommand command) {
```
Comment (peet-j) [inline java/src/main/java/com/xxAMIDOxx/xxSTACKSxx/menu/handlers/UpdateCategoryHandler.java:58]: As mentioned above. Don't copy/paste.
Use inheritance if necessary

File: java/src/main/java/com/xxAMIDOxx/xxSTACKSxx/menu/mappers/RequestToCommandMapper.java [hunk 10]
```diff
@@ -3,31 +3,54 @@
...
 
-    public static UpdateMenuCommand map(String correlationId, UUID menuId, UpdateMenuRequest r) {
+    public static UpdateMenuCommand map(String correlationId, UUID menuId,
+                                        UpdateMenuRequest r) {
         return new UpdateMenuCommand(correlationId, menuId, r.getName(),
                 r.getDescription(), r.getEnabled());
     }
 
-    public static CreateCategoryCommand map(String correlationId, UUID menuId, CreateCategoryRequest r) {
+    public static DeleteMenuCommand map(String correlationId, UUID menuId) {
```
Comment (peet-j) [inline java/src/main/java/com/xxAMIDOxx/xxSTACKSxx/menu/mappers/RequestToCommandMapper.java:32]: Unnecessary

File: java/src/test/java/com/xxAMIDOxx/xxSTACKSxx/menu/api/v1/impl/DeleteCategoryControllerImplTest.java [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+        var response =
+                this.testRestTemplate.exchange(requestUrl, HttpMethod.DELETE, requestEntity, ErrorResponse.class);
+
+        // Then
+        ArgumentCaptor<Menu> captor = ArgumentCaptor.forClass(Menu.class);
+        verify(repository, times(0)).save(menu);
+        then(response.getStatusCode()).isEqualTo(CONFLICT);
+    }
+
+    private HttpEntity<Object> getObjectHttpEntity() {
```
Comment (peet-j) [inline java/src/test/java/com/xxAMIDOxx/xxSTACKSxx/menu/api/v1/impl/DeleteCategoryControllerImplTest.java:111]: Strange place to put this. I'm guessing you've added the same thing elsewhere?
As mentioned, don't duplicate code

File: java/src/test/java/com/xxAMIDOxx/xxSTACKSxx/menu/api/v1/impl/DeleteCategoryControllerImplTest.java [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+                this.testRestTemplate.exchange(requestUrl, HttpMethod.DELETE, requestEntity, ErrorResponse.class);
+
+        // Then
+        ArgumentCaptor<Menu> captor = ArgumentCaptor.forClass(Menu.class);
+        verify(repository, times(1)).save(menu);
+        then(response.getStatusCode()).isEqualTo(OK);
+    }
+
+    @Test
+    void testDeleteCategoryWithAnItem() {
```
Comment (peet-j) [inline java/src/test/java/com/xxAMIDOxx/xxSTACKSxx/menu/api/v1/impl/DeleteCategoryControllerImplTest.java:87]: Don't think this would be a failure would it? Item(s) just get deleted.
Fair enough though. Makes a nice edge case

File: java/src/test/java/com/xxAMIDOxx/xxSTACKSxx/menu/api/v1/impl/DeleteMenuControllerImplTest.java [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+    @MockBean
+    private MenuRepository repository;
+
+    @AfterEach
+    void tearDown() {
+        repository.deleteAll();
+    }
+
+    @Test
+    void testDeleteSuccess() {
```
Comment (peet-j) [inline java/src/test/java/com/xxAMIDOxx/xxSTACKSxx/menu/api/v1/impl/DeleteMenuControllerImplTest.java:58]: Failure case?

**Review by peet-j — CHANGES_REQUESTED** [review]
See comments

File: java/src/test/java/com/xxAMIDOxx/xxSTACKSxx/menu/api/v1/impl/DeleteCategoryControllerImplTest.java (hunk 12 above)
Comment (Akrish17) [reply java/src/test/java/com/xxAMIDOxx/xxSTACKSxx/menu/api/v1/impl/DeleteCategoryControllerImplTest.java]: according to the swagger spec if a category has an item we can't delete it
